document/core/valid/modules.rst

```diff
@@ -5,6 +5,61 @@ Modules
 Furthermore, most definitions are themselves classified with a suitable type.
 
 
+.. index:: type, type index, defined type, recursive type
+   pair: abstract syntax; type
+   single: abstract syntax; type
+.. _valid-types:
+
+Types
+~~~~~
+
+The sequence of :ref:`types <syntax-type>` defined in a module is validated incrementally, yielding a suitable :ref:`context <context>`.
+
+:math:`\type^\ast`
+..................
+
+* If the sequence is empty, then:
+
+  * The :ref:`context <context>` :math:`C` must be empty.
+
+  * Then the type sequence is valid.
+
+* Otherwise:
+
+  * Let the :ref:`recursive type <syntax-rectype>` :math:`\rectype` be the last element in the sequence.
+
+  * The sequence without :math:`\rectype` must be valid for some context :math:`C'`.
+
+  * Let the :ref:`type index <syntax-typeidx>` :math:`x` be the length of :math:`C'.\CTYPES`, i.e., the first type index free in :math:`C'`.
+
+  * Let the sequence of :ref:`defined types <syntax-deftype>` :math:`\deftype^\ast` be the result :math:`\rolldt_x(\rectype)` of :ref:`rolling up <aux-roll-deftype>` into its sequence of :ref:`defined types <syntax-deftype>`.
+
+  * The :ref:`recursive type <syntax-rectype>` :math:`\rectype` must be :ref:`valid <valid-rectype>` under the context :math:`C` for :ref:`type index <syntax-typeidx>` :math:`x`.
+
+  * The current :ref:`context <context>` :math:`C` be the same as :math:`C'`, but with :math:`\deftype^\ast` appended to |CTYPES|.
+
+  * Then the type sequence is valid.
+
+.. math::
```

Comment:
probably should move each judgement to a separate line, and separate the premises of the first judgement across multiple lines